dali/pipeline/util/thread_pool.cc

```diff
@@ -174,13 +173,47 @@ void ThreadPool::ThreadMain(int thread_id, int device_id, bool set_affinity,
       lock.unlock();
     }
 
-    // Mark this thread as idle & check for complete work
-    lock.lock();
-    --active_threads_;
-    if (work_queue_.empty() && active_threads_ == 0) {
-      work_complete_ = true;
+    // The task is now complete - we can atomically decrement the number of outstanding work.
+    // If it reaches zero, we must safely notify the potential threads waiting for the work
+    // to complete.
+    // NOTE: We don't have to acquire the mutex until the number of waiting threads reaches 0.
+    if (--outstanding_work_ == 0) {
+      // We don't need to guard the modification of the atomic value with a mutex -
+      // however, we need to lock it briefly to make sure we don't have this scenario:
+      //
+      // worker                           WaitForWork
+      //
+      //                                  lock.lock()
+      //                                  return outstanding_work_ == 0  (false!)
+      // --outstanding_work == 0 (true)
+      // compleded_.notify_all()          NOT WAITING FOR compleded_ YET!!!!!!!!!!!!!
+      //                                  atomically unlock `lock` and wait for `completed_`
+      //                                                               ^^^^ deadlock
+
+
+      // The brief lock/unlock sequence avoids this scenario:
+      //
+      // worker                           WaitForWork
+      //
+      //                                  lock.lock()
+      //                                  return outstanding_work_ == 0  (false!)
+      // --outstanding_work == 0 (true)
+      // lock.lock()
+      //                                  atomically unlock `lock` and wait for `completed_`
+      //                                                               ^^^^ deadlock
```

Comment:
The deadlock should not happen then, right?

Reply:
Of course it shouldn't.